cluster-autoscaler/simulator/cluster_snapshot.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	apiv1 "k8s.io/api/core/v1"
+	schedulerlisters "k8s.io/kubernetes/pkg/scheduler/listers"
+)
+
+// ClusterSnapshot is abstraction of cluster state used for predicate simulations.
+// It exposes efficient mutation methods and can be viewed as scheduler's SharedLister
+// via GetSchedulerLister() method.
+type ClusterSnapshot interface {
```

Comment:
What do you think about adding something like `Snapshot(ListerRegistry) error` which would grab list of ready/all nodes, pods, etc. Than we could add getters for those and just start passing ClusterSnapshot rather than lists of pods/nodes to everywhere. 
Pros:
 - One source of truth for snapshot.
 - We encapsulate some stuff.
Cons:
 - Probably a lot of mechanical code changes, adds to future problems with merges.